### ISSUE TITLE
AIX DDR compile failure with new installation of 7.1.5.5

### ISSUE THREAD

**Issue by jdekonin** [eclipse/openj9#8508]
The failure also occurs on 7.2.

```
OpenJ9 compile complete
/opt/freeware/bin/gmake -s VERBOSE="-s" LOG_LEVEL="warn" -I /home/u0020236/workspace/openj9-openjdk-jdk8/make/common -f /home/u0020236/workspace/openj9-openjdk-jdk8/closed/OpenJ9.gmk SPEC=/home/u0020236/workspace/openj9-openjdk-jdk8/build/aix-ppc64-normal-server-release/spec.gmk stage_openj9_build_jdk
Running ddrgen to generate j9ddr.dat and superset.dat
gmake[2]: *** [run_omrddrgen.mk:74: ../j9ddr.dat] IOT/Abort trap (core dumped)
gmake[1]: *** [/home/u0020236/workspace/openj9-openjdk-jdk8/closed/OpenJ9.gmk:225: run-ddrgen] Error 2
make: *** [/home/u0020236/workspace/openj9-openjdk-jdk8/closed/make/Main.gmk:42: j9vm-compose-buildjvm] Error 2
```

Breaking it down to the failing cmdln:
```
LIBPATH=.. ../ddrgen --blob ../j9ddr.dat --superset ../superset.dat --blacklist blacklist --macrolist ../macroList --overrides overrides --show-empty  ../libj9ddr_misc29.so.dbg ../libj9gc29.so.dbg ../libj9jvmti29.so.dbg ../libj9prt29.so.dbg ../libj9shr29.so.dbg ../libj9thr29.so.dbg ../libj9trc29.so.dbg ../libj9vm29.so.dbg ../libj9vrb29.so.dbg ../libjclse29.so.dbg
```

`LIBPATH=.. ../ddrgen` does successfully print out ddrgen help though.

**Comment by pshipton** [comment]
This is blocking the OpenJ9 project from having more than 1 AIX machine to compile on.

**Comment by jdekonin** [comment]
I have core files for anyone that wants them.  As well as an environment to reproduce.

```IOT/Abort trap in pthread_kill at 0x900000000587f94 ($t1)
0x900000000587f94 (pthread_kill+0xd4) e8410028             ld   r2,0x28(r1)
(dbx) where
pthread_kill(??, ??) at 0x900000000587f94
_p_raise(??) at 0x9000000005877e4
raise.raise(??) at 0x9000000000397e8
abort() at 0x9000000000561e4
std::myabort()() at 0x900000000ba100c
std::terminate()() at 0x900000000ba0ee8
exceptio.std::terminate().terminate()() at 0x900000000ba190c
__DoThrowV6() at 0x900000000ba2a50
unnamed block in ddrgen.std::vector<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >,std::allocator<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> > > >::insert(std::_Ptrit<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >,long,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >*,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >&,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >*,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >&>,unsigned long,const std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >&)(??, ??, ??, ??), line 283 in "xmemory"
unnamed block in ddrgen.std::vector<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >,std::allocator<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> > > >::insert(std::_Ptrit<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >,long,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >*,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >&,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >*,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >&>,unsigned long,const std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >&)(??, ??, ??, ??), line 283 in "xmemory"
ddrgen.std::vector<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >,std::allocator<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> > > >::insert(std::_Ptrit<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >,long,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >*,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >&,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >*,std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >&>,unsigned long,const std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >&)(??, ??, ??, ??), line 283 in "xmemory"
AixSymbolTableParser.split(const std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >&,char,std::vector<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> >,std::allocator<std::_LFS_ON::basic_string<char,std::char_traits<char>,std::allocator<char> > > >*)(data = &(...), delimeters = '^O', elements = 0x00000000000001a0, _S = [internal error: nil paramlist for function containing _S]
```

**Comment by pshipton** [comment]
@jdekonin what is the version of the compiler on these machines? `xlc -qversion`
How does it compare to the version on aix71-p8-1?

**Comment by jdekonin** [comment]
```
IBM XL C/C++ for AIX, V13.1.3 (5725-C72, 5765-J07)
Version: 13.01.0003.0000
...
IBM XL C/C++ for AIX, V16.1.0  (5725-C72, 5765-J12)
Version: 16.01.0000.0000
```

This the same versions being used without issue on aix71-p8-1.  I am also retrying with the latest patch for both the compilers too.  JDK8 = xlc13, JDK14 = xlc16

**Comment by keithc-ca** [comment]
@jdekonin Can you try another build with the change in eclipse/omr#4815?

**Comment by jdekonin** [comment]
What sha of openj9 or extensions should I use with this?  I've pulled your fork and the latest for openj9 and extensions and I get compile failures in omr.

**Comment by jdekonin** [comment]
I had upgraded the compile to 16.01.0000.0004.  OMR no longer compiles with this version.

```
/opt/IBM/xlC/16.1.0/bin/xlclang  -I. -I../include -I../stats -I../structs -Istandard -Isegregated -I../../../gc_glue_java -I../../../oti -I../../../include -I../../../gc_base -I../../../gc_include -I../../../gc_stats -I../../../gc_structs -I../../../gc_base -I../../../include -I../../../oti -I../../../nls -I../../../gc_include -I../../../gc_structs -I../../../gc_stats -I../../../gc_modron_standard -I../../../gc_realtime -I../../../gc_trace -I../../../gc_vlhgc -Isegregated -Ivlhgc  -I../../../include -I../../include_core -I../../nls -DPPC64 -D_XOPEN_SOURCE_EXTENDED=1 -D_ALL_SOURCE -DRS6000 -DAIXPPC -D_LARGE_FILES -c  -s -q64 -qarch=ppc -qalias=noansi -qxflag=LTOL:LTOL0 -qsuppress=1506-1108 -qlanglvl=extended0x -qxlcompatmacros -O3 -qhalt=w -g   -o ut_omrmm.o ut_omrmm.c
error: invalid argument '-std=gnu++0x' not allowed with 'C/ObjC'
Error while processing ut_omrmm.c.
gmake[7]: *** [../../omrmakefiles/rules.mk:372: ut_omrmm.o] Error 1
```

`-qlanglvl=extended0x ` mapping to -std sub-option `c++0x | gnu++11 | gnu++0x`
https://www.ibm.com/support/knowledgecenter/SSGH3R_16.1.0/com.ibm.xlcpp161.aix.doc/compiler_ref/opt_langlvl_clang.html

But that can be resolved in a different issue (and within the Eclipse OMR project).

**Comment by DanHeidinga** [comment]
@mikezhang1234567890 Can you take a look at this once JEP 352 is complete?

**Comment by mikezhang1234567890** [comment]
Will take a look.

**Comment by sxa555** [comment]
For reference we are building on AIX `7100-04-01` at AdoptOpenJDK with xlc `16.01.0000.0000` (We did briefly run the builds on 7.1TL05SP05 but moved back for compatibility and because the OpenJ9 requirement from TL05SP05 went away for now.

**Comment by pshipton** [comment]
> because the OpenJ9 requirement from TL05SP05 went away for now

It's not exactly gone away. TL05SP05 has a properly working `dump` utility and the older AIX versions do not. However we're not aware of any problems atm caused by the old `dump`.

**Comment by jdekonin** [comment]
So it appears that setting the data seg size to 1572864 gets this working.  I had noticed that `export LDR_CNTRL=MAXDATA=0x80000000` was not actually being set with a manual logon but that apparently has no impact on this particular failure.

aix-openj9-1 differs from the documented instructions for `data seg size` and `stack size`.
data seg size           (kbytes, -d) unlimited
stack size              (kbytes, -s) 4194304

I don't have any references for why these were changed to the current settings.  I don't see any issues that lead me to believe that there was a required change.  Can I please get a clarification to what should be used for defaults.  For reference system defaults for the 2 are data(262144) and stack(65536).

**Comment by pshipton** [comment]
I know we recently made this change to set the data soft limit to the hard limit.
https://github.com/eclipse/openj9/pull/8367
If the data limit is too low you can get OutOfMemoryError trying to create a new Thread.

**Comment by jdekonin** [comment]
Problems creating a new thread would be related to `max user processes` no?

The hard limit for `data seg size` on the systems I have access to default to unlimited.  I'm not certain we want to be doing that.  @gita-omr @zl-wang would either of you have suggestions as to ulimit settings to use for our AIX machines; specifically `data seg size`.

**Comment by jdekonin** [comment]
Currentl ulimit settings are 
```
-bash-5.0$ ulimit -a
core file size          (blocks, -c) unlimited
data seg size           (kbytes, -d) unlimited
file size               (blocks, -f) unlimited
max memory size         (kbytes, -m) 65536
open files                      (-n) unlimited
pipe size            (512 bytes, -p) 64
stack size              (kbytes, -s) 65536
cpu time               (seconds, -t) unlimited
max user processes              (-u) unlimited
virtual memory          (kbytes, -v) unlimited
```

Not so sure using ulimited as values for some of these is a good idea, I am looking for suggestions as to what would be expected or common settings to use.

**Comment by zl-wang** [comment]
-m setting is too low, while others are ok.  i would suggest -m going to unlimited, good for all possible test cases.

**Comment by sxa** [comment]
@zi-wang In my view the purpose shouldn't be to set unlimited for everything "just to make the tests pass"  as it can mask leaks and makes it harder to provide adequate suggestions for end-users as to when they might need to increase them.

**Comment by zl-wang** [comment]
@sxa  that is true in theory.  of course, setting it to unlimited is not for the purpose of "passing DDR tests".  depending on limits to catch possible leaks is not a good strategy, or might not be as useful as you think.  in reality, every environment (AIX/Linux/iOS) i have seen by default set -m to unlimited already, while some other limits (in particular -c/-n/-p/-s) are with specific numbers.

**Comment by sxa** [comment]
If it's a default, that's completely ok. But if it's not then we should not be automatically overriding it.

**Comment by keithc-ca** [comment]
@jdekonin Is anything left to be done here?

**Comment by jdekonin** [comment]
I never got an answer to my question about what we should be using for `data seg size`.  I started searching for documentation, and I found that IBM Java documentation details that -c, -n, -d, -f all need to be set to unlimited in order to enable full core dump generation.
https://www.ibm.com/support/knowledgecenter/SSYKE2_8.0.0/com.ibm.java.vm.80.doc/docs/j9_configure_ulimits.html

So i guess there is nothing left to be done here.
